### ISSUE TITLE
Reprocibility

### ISSUE THREAD

**Issue by jadkins99** [danijar/dreamerv3#43]
Awesome repo. quick question, 
[runs.json.gz](https://github.com/danijar/dreamerv3/files/11240742/runs.json.gz)
I ran the DMC WalkerWalk experiment 3 different times with the same seeds and got 3 different learning curves.  How can I get reproducible experiments?Awesome repo. quick question, 
[runs.json.gz](https://github.com/danijar/dreamerv3/files/11240742/runs.json.gz)
I ran the DMC WalkerWalk experiment 3 different times with the same seeds and got 3 different learning curves.  How can I get reproducible experiments?
![curves](https://user-images.githubusercontent.com/48653578/232254643-bc9e29c4-e9b6-41d4-b9ee-646a255a4560.png)
![curves](https://user-images.githubusercontent.com/48653578/232254648-43265f09-4098-4cef-8496-aecbfe47bf68.png)
![curves](https://user-images.githubusercontent.com/48653578/232254659-fb29ceae-cec9-40fa-b947-0fa43f9e91c9.png)

**Comment by danijar** [comment]
Hi, are you asking for fully deterministic runs? I haven't paid much attention to this but I think the agent is already fully deterministic, so you'd probably just have to set the environment seed (make sure if you use more than 1 environment instance, that the environments have different seeds so they produce different data).

**Comment by jadkins99** [comment]
Okay I will try that. Thank you for the quick response! What exactly is an "environment instance"? I couldn't find a clear definition in the paper.

**Comment by jadkins99** [comment]
Also, how many seeds were the non-Minecraft experiments run for?

**Comment by subho406** [comment]
+1 on the question above. Maybe it's not that apparent in the paper,  could you also provide some clarification on what the confidence intervals denote in the non-minecraft experiments (DMLab, DMC Proprio, Crafter, etc)? Is it std-error multiple seeds, std-error across a window of timesteps with a single seed, or something else?

**Comment by danijar** [comment]
It's mean/std across seeds and at least 3 seeds per task, often more.

**Comment by jadkins99** [comment]
Update: I seeded  [dmc_control ](https://github.com/danijar/dreamerv3/blob/84ecf191d967f787f5cc36298e69974854b0df9c/dreamerv3/embodied/envs/dmc.py#L34) here. And still got non-deterministic runs.

**Comment by jadkins99** [comment]
I found some non-seeded randomness in the repo. Namely [here](https://github.com/danijar/dreamerv3/blob/de7ad471d93c8bdb6140551984ca2b53c3add657/dreamerv3/behaviors.py#L50) and [here](https://github.com/danijar/dreamerv3/blob/de7ad471d93c8bdb6140551984ca2b53c3add657/dreamerv3/embodied/core/wrappers.py#L274) and [here](https://github.com/danijar/dreamerv3/blob/de7ad471d93c8bdb6140551984ca2b53c3add657/dreamerv3/embodied/core/space.py#L17). Wouldn't these affect the agent?

**Comment by jadkins99** [comment]
Also, If run.scripts == 'train', there should be no multiprocessing correct? I am a bit confused about what happens when num_envs > 1 and run.scripts != 'parallel'.

**Comment by danijar** [comment]
I don't think those two methods are run ever. Could you check e.g. by adding `asdf` to the two methods to see if it errors?

**Comment by swannercjj** [comment]
Seeding [this](https://github.com/danijar/dreamerv3/blob/de7ad471d93c8bdb6140551984ca2b53c3add657/dreamerv3/embodied/core/space.py#L17) it removes randomness from the first 1000 steps, but runs are non-deterministic afterwards.

**Comment by Dongyeongkim** [comment]
@jadkins99  @swannercjj  I guess you should check this issue: https://github.com/google/jax/issues/13672#issuecomment-1515544978 

It seems just seeding on jax.random does not make it as deterministic operation on GPU because of the optimisation in graph compilation process.
